src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/RealWorldContextTests.cs

```diff
@@ -882,5 +882,121 @@ public virtual void NullableStruct()
+
+        [Fact]
+        public void JsonIgnoreForCovariantProperties()
+        {
+            var derived = new CovariantDerived();
+
+            if (DefaultContext.JsonSourceGenerationMode == JsonSourceGenerationMode.Serialization)
+            {
+                Assert.Throws<InvalidOperationException>(() => JsonSerializer.Serialize(derived, DefaultContext.CovariantDerived));
+            }
+            else
+            {
+                string json = JsonSerializer.Serialize(derived, DefaultContext.CovariantDerived);
+                JsonTestHelper.AssertJsonEqual(@"{}", json);
+                derived = JsonSerializer.Deserialize(json, DefaultContext.CovariantDerived);
+                Assert.Null(derived.Id);
+            }
+        }
+
+        [Fact]
+        public void JsonIgnoreForCovariantGenericProperties()
+        {
+            var derived = new CovariantDerivedGeneric<string>();
+            if (DefaultContext.JsonSourceGenerationMode == JsonSourceGenerationMode.Serialization)
+            {
+                Assert.Throws<InvalidOperationException>(() => JsonSerializer.Serialize(derived, DefaultContext.CovariantDerivedGenericString));
+            }
+            else
+            {
+                string json = JsonSerializer.Serialize(derived, DefaultContext.CovariantDerivedGenericString);
+                JsonTestHelper.AssertJsonEqual(@"{}", json);
+
+                derived = JsonSerializer.Deserialize(json, DefaultContext.CovariantDerivedGenericString);
+                Assert.Null(derived.Id);
+            }
+        }
+
+        [Fact]
+        public void JsonIgnoreForHiddenProperties_IgnoredBase_NotIgnoredDerived()
+        {
+            var derived = new IgnoredPropertyBase_NotIgnoredPropertyDerived { Id = "Test" };
+
+            string json = JsonSerializer.Serialize(derived, DefaultContext.IgnoredPropertyBase_NotIgnoredPropertyDerived);
+            JsonTestHelper.AssertJsonEqual("{\"Id\":\"Test\"}", json);
+
+            if (DefaultContext.JsonSourceGenerationMode == JsonSourceGenerationMode.Serialization)
+            {
+                // Deserialization not supported in fast path serialization only mode
+                Assert.Throws<InvalidOperationException>(() => JsonSerializer.Deserialize(json, DefaultContext.IgnoredPropertyBase_NotIgnoredPropertyDerived));
+            }
+            else
+            {
+                derived = JsonSerializer.Deserialize(json, DefaultContext.IgnoredPropertyBase_NotIgnoredPropertyDerived);
+                Assert.Equal("Test", derived.Id);
+            }
+        }
+
+        [Fact]
+        public void JsonIgnoreForHiddenProperties_NotIgnoredBase_IgnoredDerived()
+        {
+            var derived = new NotIgnoredPropertyBase_IgnoredPropertyDerived { Id = "Test" };
+
+            string json = JsonSerializer.Serialize(derived, DefaultContext.NotIgnoredPropertyBase_IgnoredPropertyDerived);
+            JsonTestHelper.AssertJsonEqual("{\"Id\":null}", json);
```

Comment:
I'm not sure this should be expected, considering that for `JsonIgnoreForHiddenProperties_NotIgnoredBase_NotIgnoredDerived` we're not showing both `Id`s my gut tells me I should not be seeing any `Id`s here either. We should be consistent for hidden non-ignored properties I think. cc: @layomia @eiriktsarpalis @steveharter

Reply:
I'm not sure either, what are the semantics in the reflection-based for that scenario?

Reply:
I tried it out without JsonSerializerContext with the following code:

```csharp
using System.Text.Json;
using System.Text.Json.Serialization;

NotIgnoredBase b1 = new NotIgnoredBase();
NotIgnoredBase_NotIgnoredDerived b2 = new NotIgnoredBase_NotIgnoredDerived();
NotIgnoredBase b3 = new NotIgnoredBase_NotIgnoredDerived();
NotIgnoredBase_IgnoredDerived b4 = new NotIgnoredBase_IgnoredDerived();
NotIgnoredBase b5 = new NotIgnoredBase_IgnoredDerived();
b1.Property = "1"; // Property on base
b2.Property = "2";
b3.Property = "3"; // Property on base
b4.Property = "4";
b5.Property = "5"; // Property on base

Console.WriteLine(JsonSerializer.Serialize(b1));
Console.WriteLine(JsonSerializer.Serialize(b2));
Console.WriteLine(JsonSerializer.Serialize(b3));
Console.WriteLine(JsonSerializer.Serialize(b4));
Console.WriteLine(JsonSerializer.Serialize(b5));

Console.WriteLine();

Console.WriteLine(JsonSerializer.Serialize(b1, typeof(NotIgnoredBase)));
Console.WriteLine(JsonSerializer.Serialize(b2, typeof(NotIgnoredBase)));
Console.WriteLine(JsonSerializer.Serialize(b2, typeof(NotIgnoredBase_NotIgnoredDerived)));
Console.WriteLine(JsonSerializer.Serialize(b3, typeof(NotIgnoredBase)));
Console.WriteLine(JsonSerializer.Serialize(b4, typeof(NotIgnoredBase)));
Console.WriteLine(JsonSerializer.Serialize(b4, typeof(NotIgnoredBase_IgnoredDerived)));
Console.WriteLine(JsonSerializer.Serialize(b5, typeof(NotIgnoredBase)));

public class NotIgnoredBase
{
	public string Property { get; set; } = nameof(NotIgnoredBase);
}
public class NotIgnoredBase_NotIgnoredDerived : NotIgnoredBase
{
	public new string Property { get; set; } = nameof(NotIgnoredBase_NotIgnoredDerived);
}
public class NotIgnoredBase_IgnoredDerived : NotIgnoredBase
{
	[JsonIgnore]
	public new string Property { get; set; } = nameof(NotIgnoredBase_IgnoredDerived);
}
```

Results:

```json
{"Property":"1"}
{"Property":"2"}
{"Property":"3"}
{"Property":"NotIgnoredBase"}
{"Property":"5"}

{"Property":"1"}
{"Property":"NotIgnoredBase"}
{"Property":"2"}
{"Property":"3"}
{"Property":"NotIgnoredBase"}
{"Property":"NotIgnoredBase"}
{"Property":"5"}
```

Reply:
It turns out the reflection-based serializer also throws an exception if the properties are ignored both on the base and the derived class.

```csharp
	IgnoredBase_IgnoredDerived b4 = new IgnoredBase_IgnoredDerived();
	Console.WriteLine(JsonSerializer.Serialize(b4));


public class IgnoredBase
{
	[JsonIgnore]
	public string Property { get; set; } = nameof(IgnoredBase);
}
public class IgnoredBase_IgnoredDerived : IgnoredBase
{
	[JsonIgnore]
	public new string Property { get; set; } = nameof(IgnoredBase_IgnoredDerived);
}
```
Exception:
```
System.ArgumentException: An item with the same key has already been added. Key: Property
   at System.ThrowHelper.ThrowAddingDuplicateWithKeyArgumentException[T](T key)
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo.CacheMember(JsonPropertyInfo jsonPropertyInfo, JsonPropertyDictionary`1 propertyCache, Dictionary`2& ignoredMembers) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs:line 409
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo.CacheMember(Type declaringType, Type memberType, MemberInfo memberInfo, Boolean isVirtual, Nullable`1 typeNumberHandling, Boolean& propertyOrderSpecified, Dictionary`2& ignoredMembers) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs:line 373
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo..ctor(Type type, JsonConverter converter, Type runtimeType, JsonSerializerOptions options) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs:line 224
   at System.Text.Json.JsonSerializerOptions.<InitializeForReflectionSerializer>g__CreateJsonTypeInfo|112_0(Type type, JsonSerializerOptions options) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:line 596
   at System.Text.Json.JsonSerializerOptions.GetClassFromContextOrCreate(Type type) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:line 625
   at System.Text.Json.JsonSerializerOptions.GetOrAddClass(Type type) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:line 605
   at System.Text.Json.JsonSerializer.GetTypeInfo(JsonSerializerOptions options, Type runtimeType) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Helpers.cs:line 26
   at System.Text.Json.JsonSerializer.Serialize(Object value, Type inputType, JsonSerializerOptions options) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.String.cs:line 63
   at Program.<Main>$(String[] args) in C:\src\JsonTest\JsonTest\Program.cs:line 54
```

This error would be also fixed by the PR:
https://github.com/dotnet/runtime/pull/63823/files#diff-8d841bf86f2a9da83758276bd17f55d002ee5b55f58b72da658df351a7c22bc0L412-R412

Reply:
Full code to compare the source generator and reflection-based output:

```csharp
using System;
using System.Text.Json;
using System.Text.Json.Serialization;
{
	NotIgnoredBase b1 = new NotIgnoredBase();
	NotIgnoredBase_NotIgnoredDerived b2 = new NotIgnoredBase_NotIgnoredDerived();
	NotIgnoredBase b3 = new NotIgnoredBase_NotIgnoredDerived();
	NotIgnoredBase_IgnoredDerived b4 = new NotIgnoredBase_IgnoredDerived();
	NotIgnoredBase b5 = new NotIgnoredBase_IgnoredDerived();
	b1.Property = "1"; // Property on base
	b2.Property = "2";
	b3.Property = "3"; // Property on base
	b4.Property = "4";
	b5.Property = "5"; // Property on base

	Console.WriteLine("NotIgnoredBase, Reflection-based:");
	Console.WriteLine(JsonSerializer.Serialize(b1, typeof(NotIgnoredBase)));
	Console.WriteLine(JsonSerializer.Serialize(b2, typeof(NotIgnoredBase)));
	Console.WriteLine(JsonSerializer.Serialize(b2, typeof(NotIgnoredBase_NotIgnoredDerived)));
	Console.WriteLine(JsonSerializer.Serialize(b3, typeof(NotIgnoredBase)));
	Console.WriteLine(JsonSerializer.Serialize(b4, typeof(NotIgnoredBase)));
	Console.WriteLine(JsonSerializer.Serialize(b4));
	Console.WriteLine(JsonSerializer.Serialize(b5, typeof(NotIgnoredBase)));

	Console.WriteLine();
	Console.WriteLine("NotIgnoredBase, Source generator-based:");
	Console.WriteLine(JsonSerializer.Serialize(b1, JsonContext.Default.NotIgnoredBase));
	Console.WriteLine(JsonSerializer.Serialize(b2, JsonContext.Default.NotIgnoredBase));
	Console.WriteLine(JsonSerializer.Serialize(b2, JsonContext.Default.NotIgnoredBase_NotIgnoredDerived));
	Console.WriteLine(JsonSerializer.Serialize(b3, JsonContext.Default.NotIgnoredBase));
	Console.WriteLine(JsonSerializer.Serialize(b4, JsonContext.Default.NotIgnoredBase));
	Console.WriteLine(JsonSerializer.Serialize(b4, JsonContext.Default.NotIgnoredBase_IgnoredDerived));
	Console.WriteLine(JsonSerializer.Serialize(b5, JsonContext.Default.NotIgnoredBase));

}
{
	IgnoredBase b1 = new IgnoredBase();
	IgnoredBase_NotIgnoredDerived b2 = new IgnoredBase_NotIgnoredDerived();
	IgnoredBase b3 = new IgnoredBase_NotIgnoredDerived();
	IgnoredBase_IgnoredDerived b4 = new IgnoredBase_IgnoredDerived();
	IgnoredBase b5 = new IgnoredBase_IgnoredDerived();
	b1.Property = "1"; // Property on base
	b2.Property = "2";
	b3.Property = "3"; // Property on base
	b4.Property = "4";
	b5.Property = "5"; // Property on base
	Console.WriteLine();
	Console.WriteLine("IgnoredBase, Reflection-based:");
	Console.WriteLine(JsonSerializer.Serialize(b1, typeof(IgnoredBase)));
	Console.WriteLine(JsonSerializer.Serialize(b2, typeof(IgnoredBase)));
	Console.WriteLine(JsonSerializer.Serialize(b2, typeof(IgnoredBase_NotIgnoredDerived)));
	Console.WriteLine(JsonSerializer.Serialize(b3, typeof(IgnoredBase)));
	Console.WriteLine(JsonSerializer.Serialize(b4, typeof(IgnoredBase)));
	Console.WriteLine("Serialization error"/*JsonSerializer.Serialize(b4, typeof(IgnoredBase_IgnoredDerived))*/);
	Console.WriteLine(JsonSerializer.Serialize(b5, typeof(IgnoredBase)));

	Console.WriteLine();
	Console.WriteLine("IgnoredBase, Source generator-based:");
	Console.WriteLine(JsonSerializer.Serialize(b1, JsonContext.Default.IgnoredBase));
	Console.WriteLine(JsonSerializer.Serialize(b2, JsonContext.Default.IgnoredBase));
	Console.WriteLine(JsonSerializer.Serialize(b2, JsonContext.Default.IgnoredBase_NotIgnoredDerived));
	Console.WriteLine(JsonSerializer.Serialize(b3, JsonContext.Default.IgnoredBase));
	Console.WriteLine(JsonSerializer.Serialize(b4, JsonContext.Default.IgnoredBase));
	Console.WriteLine("Source generator error"/*JsonSerializer.Serialize(b4, JsonContext.Default.IgnoredBase_IgnoredDerived)*/);
	Console.WriteLine(JsonSerializer.Serialize(b5, JsonContext.Default.IgnoredBase));

	Console.WriteLine();
}

public class NotIgnoredBase
{
	public string Property { get; set; } = nameof(NotIgnoredBase);
}
public class NotIgnoredBase_NotIgnoredDerived : NotIgnoredBase
{
	public new string Property { get; set; } = nameof(NotIgnoredBase_NotIgnoredDerived);
}
public class NotIgnoredBase_IgnoredDerived : NotIgnoredBase
{
	[JsonIgnore]
	public new string Property { get; set; } = nameof(NotIgnoredBase_IgnoredDerived);
}
public class IgnoredBase
{
	[JsonIgnore]
	public string Property { get; set; } = nameof(IgnoredBase);
}
public class IgnoredBase_NotIgnoredDerived : IgnoredBase
{
	public new string Property { get; set; } = nameof(IgnoredBase_NotIgnoredDerived);
}
public class IgnoredBase_IgnoredDerived : IgnoredBase
{
	[JsonIgnore]
	public new string Property { get; set; } = nameof(IgnoredBase_IgnoredDerived);
}

[JsonSerializable(typeof(NotIgnoredBase))]
[JsonSerializable(typeof(NotIgnoredBase_NotIgnoredDerived))]
[JsonSerializable(typeof(NotIgnoredBase_IgnoredDerived))]
[JsonSerializable(typeof(IgnoredBase))]
[JsonSerializable(typeof(IgnoredBase_NotIgnoredDerived))]
//[JsonSerializable(typeof(IgnoredBase_IgnoredDerived))]
public partial class JsonContext : JsonSerializerContext
{ }
```

Output:
```json
NotIgnoredBase, Reflection-based:
{"Property":"1"}
{"Property":"NotIgnoredBase"}
{"Property":"2"}
{"Property":"3"}
{"Property":"NotIgnoredBase"}
{"Property":"NotIgnoredBase"}
{"Property":"5"}

NotIgnoredBase, Source generator-based:
{"Property":"1"}
{"Property":"NotIgnoredBase"}
{"Property":"2"}
{"Property":"3"}
{"Property":"NotIgnoredBase"}
{"Property":"NotIgnoredBase"}
{"Property":"5"}

IgnoredBase, Reflection-based:
{}
{}
{"Property":"2"}
{}
{}
Serialization error
{}

IgnoredBase, Source generator-based:
{}
{}
{"Property":"2"}
{}
{}
Source generator error
{}
```